### DISCUSSION TITLE
input.harbor with video playlist gives "buffer overrun"

### DISCUSSION THREAD

**Discussion by Mikulskii** [savonet/liquidsoap#3106] [Q&A] [answered]
Hello!
As a source, I use a playlist with video files and output them to rtmp.
I would like to connect sometimes as a host to this stream, but when I connecting via input.harbor, the error "buffer overrun" crawls endlessly in the log. At the same time, the sound does not come to the stream at all, even clicks or other glitches
Increasing the buffer does not help (buffer=, max=)
```
videos = playlist("path/to/mp4")
mic = mksafe(input.harbor("mount", port=, password=""))
stream = add([mic, videos])
url = ""
enc = %ffmpeg(..)
output.url(fallible=true, url=url, enc, stream)
```
At the same time, everything works fine if the audio and video sources are separate (with mux_video).

I also tried using rtmp as input - in this case everything works, but when I disable the incoming rtmp, the main stream freezes.

In the end, I launched another instance of liquidsoap, into which I input the stream twice and mixed them. In this case, everything works as expected, but there is a desynchronization of audio and video.
```
stream_video = mksafe(input.rtmp(id="video",listen=false, "rtmp://IP/stream/live"))
stream_audio = mksafe(input.rtmp(id="audio",listen=false, "rtmp://IP/stream/live"))

mic=mksafe(input.harbor("mount", port=, password="", buffer=1.5, max=3.0))
host = add(normalize=false, [mic, stream_audio])

radio = mux_audio(audio=drop_video(host), drop_audio(stream_video))
```
Attempts to split one source into video and audio give the error "a source cannot belong to two clocks".
Is it even possible to do this (to connect via harbor to videostream)?
I read in the blog and as far as I understood that in version 2.2.x it will be possible to pull tracks from the source separately, but will this solve the problem of buffer, synchronization and "two clocks"?
Any ideas how to do this on version 2.1.4?
Thank you.

**Comment by toots** [comment, marked as answer]
Hi!

When sending to a `output.url`, I believe that the output will drive latency control. Meanwhile, it is possible that the client sending data sends too much data compared to the rate at which the output sends its data, indeed.

Is the `mic` source supposed to be audio only? I this case, yes, I would suggest to use `rolling-release-v2.2.x` if only because it is where we would like to prioritize bug fixes at the moment.

With `2.2.x`, your base script would look like this:
```liquidsoap
videos = playlist("path/to/mp4")

let {audio = video_audio, ...tracks} = source.tracks(videos)

mic = input.harbor("mount", port=, password="")
let {audio = mic_audio} = source.tracks(mic)

audio = track.audio.add([mic_audio, video_audio])

stream = source(tracks.{
  audio = audio
})

url = ""
enc = %ffmpeg(..)
output.url(fallible=true, url=url, enc, stream)
```

**Comment by Mikulskii** [comment]
Yay! I quickly tested it on the Windows build and it looks like everything is working fine!
Well, I will wait for the stable version to put it on Ubuntu Vps.
Thank you so much for all your work! <3

Also, I don't know if this is a bug or my local Windows problem (there is no such thing on 2.1.4), but if:
```
s = mksafe(input.rtmp(listen=false, "rtmp://"))
url = "rtmp://"
enc = %ffmpeg(..)
output.url(fallible=true, url=url, enc, s)
```
There is a error-loop:
```
[decoder.ffmpeg:3] Requested content-type for "rtmp://": {audio=pcm(stereo),video=canvas}
[decoder.ffmpeg:3] FFmpeg recognizes "rtmp://" as video: {codec: h264, 1280x720, yuv420p}, audio: {codec: aac, 44100Hz,
 2 channel(s)}
[decoder.ffmpeg:3] Decoded content-type for "rtmp://": {audio=pcm(stereo),video=canvas}
[video:2] Feeding failed: File "src/core/stream/content_base.ml", line 285, characters 4-10: Assertion failed
```
If output is .ao all is fine

**Reply by toots** [reply]
Thanks @Mikulskii. This looks like a bug. I haven't been able to reproduce yet. Would be able to send the whole stack trace? Also perhaps the rtmp url, in private if needed (toots@rastageeks.org)

**Reply by Mikulskii** [reply]
Sure! Sent from info@mikulski.rocks

**Reply by toots** [reply]
Thanks! I was able to reproduce and fix. Just pushed the fix to `main` and `rolling-release-v2.2.x`!

**Reply by Mikulskii** [reply]
Hey @toots! Thank you for your quick reaction!
But a new problem has arisen in this case with `input.rtmp`: now it is impossible to connect to the `output.url` stream.
I use VLC player and it shows endless loading.
When the script is stopped, the message `"Waiting for main threads to terminate"`  hangs for a long time and then `"Segmentation fault (core dumped)"`.
Tested on Windows and Ubuntu VM (on Ubuntu, Installation via git clone/ opam pin).
